script/modelling/facility_location.jl

```diff
@@ -0,0 +1,262 @@
+
+using JuMP
+import GLPK
+using Plots
+
+#' ## Uncapacitated facility location
+
+#' ### Problem description
+#' 
+#' We are given
+#' * $M=\{1, \dots, m\}$ clients
+#' * $N=\{ 1, \dots, n\}$ sites where a facility can be built
+#' 
+#' **Decision variables**
+#' Decision variables are split into two categories:
+#' * Binary variable $y_{j}$ indicates whether facility $j$ is built or not
+#' * Binary variable $x_{i, j}$ indicates whether client $i$ is assigned to facility $j$
+#' 
+#' **Objective**
+#' The objective is to minimize the total cost of serving all clients.
+#' This costs breaks down into two components:
+#' * Fixed cost of building a facility.
+#' In this example, this cost is $f_{j} = 1, \ \forall j$.
+#' * Cost of serving clients from the assigned facility.
+#' In this example, the cost $c_{i, j}$
+#' of serving client $i$ from facility $j$
+#' is the euclidean distance between the two.
+#' 
+#' **Constraints**
+#' * Each customer must be served by exactly one facility
+#' * A facility cannot serve any client unless it is open
+
+#' ### MILP formulation
+#' 
+#' The problem can be formulated as the following MILP:
+#' 
+#' \begin{align}
+#' \min_{x, y} \ \ \ &
+#' \sum_{i, j} c_{i, j} x_{i, j} + 
+#' \sum_{j} f_{j} y_{j}\\
+#' s.t. &
+#' \sum_{j} x_{i, j} = 1, && \forall i \in M\\
+#' & x_{i, j} \leq y_{j}, && \forall i \in M, j \in N\\
+#' & x_{i, j}, y_{j} \in \{0, 1\}, && \forall i \in M, j \in N
+#' \end{align}
+#' 
+#' where the first set of constraints ensures
+#' that each client is served exactly once,
+#' and the second set of constraints ensures
+#' that no client is served from an unopened facility.
+
+#' ### Problem data
+
+#+
+Random.seed!(314)
+
+m = 12  # number of clients
+n = 5  # number of facility locations
+
+# Clients' locations
+Xc = rand(m)
+Yc = rand(m)
+
+# Facilities' potential locations
+Xf = rand(n)
+Yf = rand(n)
+
+# Fixed costs
+f = ones(n);
+
+# Distance
+c = zeros(m, n)
+for i in 1:m
+    for j in 1:n
+        c[i, j] = norm([Xc[i] - Xf[j], Yc[i] - Yf[j]], 2)
+    end
+end
+
+#+
+# Display the data
+scatter(Xc, Yc, label = "Clients", markershape=:circle, markercolor=:blue)
+scatter!(Xf, Yf, label="Facility", 
+    markershape=:square, markercolor=:white, markersize=6,
+    markerstrokecolor=:red, markerstrokewidth=2
+)
+
+#' ### JuMP implementation
+
+#+
+# Create a JuMP model
+ufl = Model(GLPK.Optimizer)
+
+# Variables
+@variable(ufl, y[1:n], Bin);
+@variable(ufl, x[1:m, 1:n], Bin);
+
+# Each client is served exactly once
+@constraint(ufl, client_service[i in 1:m],
+    sum(x[i, j] for j in 1:n) == 1
+);
+
+# A facility must be open to serve a client
+@constraint(ufl, open_facility[i in 1:m, j in 1:n],
+    x[i, j] <= y[j]
+)
+
+# Objective
+@objective(ufl, Min, f'y + sum(c .* x));
+
+#+
+# Solve the uncapacitated facility location problem with GLPK
+optimize!(ufl)
+println("Optimal value: ", objective_value(ufl))
+
+
+#' ### Visualizing the solution
+
+#+
+x_ = value.(x) .> 1 - 1e-4
```

Comment:
I would maybe add a short sentence to explain to newcomers why we are putting a threshold `1e-4` here